File: programming/Programmers/c++/프린터.cpp
/*
프로그래머스 프린터
queue 2개와 priority_queue 1개를 사용하여 해결
queue에는 각각 우선 순위와 위치를 삽입
priority_queue에는 우선 순위를 삽입
반복문을 통해 문제의 조건을 만들어 비교
*/
#include <string>
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

int solution(vector<int> priorities, int location) {
	int answer = 0;
	queue<int> q1, q2;
	priority_queue<int, vector<int>, less<int>> pq;
	for (int i = 0; i<priorities.size(); i++) {
		q1.push(priorities[i]);
		q2.push(i);
		pq.push(priorities[i]);
	}
	while (1) {
		if (q1.front() == pq.top() && q2.front() == location) break; // 찾는 문서
		else if (q1.front() == pq.top()) { // 먼저 빠져나가는 문서
			pq.pop();
			q1.pop();
			q2.pop();
			answer++;
		}
		else { // 빠져나가지 못하는 문서
			q1.push(q1.front());
			q1.pop();
			q2.push(q2.front());
			q2.pop();
		}
	}
	return answer + 1;
}

File: Programmers/javascript/프린터.js
/*
프로그래머스 프린터
작업의 순서가 있는 배열을 index와 함께 다른 배열에 저장
기존 우선순위 배열은 내림차순 정렬
우선순위가 같고 index가 location과 같다면 정지
우선순위만 같다면 각각의 배열의 다음값 호출
모두 틀리다면 복사한 배열의 다음값 호출
*/
function solution(priorities, location) {
    var answer = 0;
    var arr=[];
    for(let i=0; i<priorities.length; i++){
        arr.push([priorities[i],i]);
    }
    priorities.sort(function(a,b){
        return b-a;
    });
    let i=0;
    let j=0;
    while(1){
        if(priorities[j]==arr[i%priorities.length][0]){
            if(arr[i%priorities.length][1]==location){
                answer=j;
                break;
            }
            else {
                j++;
                i++;
            }
        }
        else i++;
    }
    //console.log(arr);
    return answer+1;
}
